File: tooling/harness-eval/sources.ts
import {
  type BehavioralCase,
  type Fixture,
  type Trigger,
  caseSchema,
  casesSchema,
  fixtureSchema,
  pathSchema,
  triggerSchema,
} from "./contracts.ts";
import { dirname, relative, resolve, sep } from "node:path";
import { readFileSync, realpathSync } from "node:fs";
import { createHash } from "node:crypto";

const fingerprint = (value: string): string =>
  createHash("sha256").update(value).digest("hex");

function sourcePath(root: string, path: string): string {
  pathSchema.parse(path);
  const canonical = realpathSync(resolve(root, path));
  const child = relative(realpathSync(root), canonical);
  if (child.startsWith(`..${sep}`) || child === ".." || child.startsWith(sep)) {
    throw new Error(`Escaping source: ${path}`);
  }
  return canonical;
}

function readSource(root: string, path: string): string {
  return readFileSync(sourcePath(root, path), "utf8");
}

function section(markdown: string, heading: string): string {
  const start = markdown.indexOf(`## ${heading}\n`);
  if (start === -1) {
    throw new Error(`Missing section: ${heading}`);
  }
  const end = markdown.indexOf("\n## ", start + 1);
  return markdown.slice(start, end === -1 ? undefined : end);
}

function loadTrigger(root: string, path: string): Trigger {
  const trigger = triggerSchema.parse(JSON.parse(readSource(root, path)));
  if (trigger.skill !== dirname(dirname(path)).split("/").at(-1)) {
    throw new Error(`Skill slug mismatch: ${path}`);
  }
  return trigger;
}

function loadCases(root: string): readonly LoadedCase[] {
  const definitions: readonly BehavioralCase[] = casesSchema.parse(
    JSON.parse(readSource(root, "harness/evals/cases.json")),
  );
  return definitions.map((definition) => resolveCase(root, definition));
}

function resolveCase(root: string, raw: unknown): LoadedCase {
  const definition = caseSchema.parse(raw);
  const sources = definition.sources.map((source) => {
    const content = readSource(root, source.path);
    section(content, source.heading);
    return { ...source, fingerprint: fingerprint(content) };
  });
  const prompt =
    "text" in definition.prompt
      ? definition.prompt.text
      : loadTrigger(root, definition.prompt.triggerFile).queries[
          definition.prompt.queryIndex
        ]?.query;
  if (prompt === undefined || prompt === "") {
    throw new Error(`Unresolved prompt: ${definition.id}`);
  }
  const fixture = fixtureSchema.parse(
    JSON.parse(
      readSource(root, `harness/evals/fixtures/${definition.fixture}.json`),
    ),
  );
  if (fixture.id !== definition.fixture) {
    throw new Error(`Fixture identity mismatch: ${definition.id}`);
  }
  return {
    definition,
    sources,
    prompt,
    promptFingerprint: fingerprint(prompt),
    fixture,
  };
}

type LoadedCase = Readonly<{
  definition: BehavioralCase;
  sources: readonly Readonly<{
    path: string;
    heading: string;
    fingerprint: string;
  }>[];
  prompt: string;
  promptFingerprint: string;
  fixture: Fixture;
}>;

export {
  fingerprint,
  sourcePath,
  readSource,
  section,
  loadTrigger,
  loadCases,
  type LoadedCase,
};

File: tooling/arnes/src/eval/sources.rs
use super::contracts::{BehavioralCase, Fixture, Prompt, Trigger, validate_path};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{
    collections::HashSet,
    fs,
    path::{Path, PathBuf},
};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ResolvedSource {
    pub path: String,
    pub heading: String,
    pub fingerprint: String,
}

#[derive(Clone, Debug)]
pub struct LoadedCase {
    pub definition: BehavioralCase,
    pub sources: Vec<ResolvedSource>,
    pub prompt: String,
    pub prompt_fingerprint: String,
    pub fixture: Fixture,
}

pub fn fingerprint(value: impl AsRef<[u8]>) -> String {
    format!("{:x}", Sha256::digest(value.as_ref()))
}

pub fn source_path(root: &Path, path: &str) -> Result<PathBuf, String> {
    validate_path(path)?;
    let root = root.canonicalize().map_err(|e| e.to_string())?;
    let canonical = root.join(path).canonicalize().map_err(|e| e.to_string())?;
    if !canonical.starts_with(&root) {
        return Err(format!("Escaping source: {path}"));
    }
    Ok(canonical)
}

pub fn read_source(root: &Path, path: &str) -> Result<String, String> {
    fs::read_to_string(source_path(root, path)?).map_err(|e| e.to_string())
}

pub fn section<'a>(markdown: &'a str, heading: &str) -> Result<&'a str, String> {
    let start = markdown
        .find(&format!("## {heading}\n"))
        .ok_or_else(|| format!("Missing section: {heading}"))?;
    let end = markdown[start + 1..]
        .find("\n## ")
        .map_or(markdown.len(), |i| start + 1 + i);
    Ok(&markdown[start..end])
}

pub fn load_trigger(root: &Path, path: &str) -> Result<Trigger, String> {
    let trigger: Trigger =
        serde_json::from_str(&read_source(root, path)?).map_err(|e| e.to_string())?;
    trigger.validate()?;
    let slug = Path::new(path)
        .parent()
        .and_then(Path::parent)
        .and_then(Path::file_name)
        .and_then(|s| s.to_str());
    if slug != Some(trigger.skill.as_str()) {
        return Err(format!("Skill slug mismatch: {path}"));
    }
    Ok(trigger)
}

pub fn load_cases(root: &Path) -> Result<Vec<LoadedCase>, String> {
    let definitions: Vec<BehavioralCase> =
        serde_json::from_str(&read_source(root, "harness/evals/cases.json")?)
            .map_err(|e| e.to_string())?;
    if definitions.is_empty() {
        return Err("Empty cases".into());
    }
    let mut ids = HashSet::new();
    definitions
        .into_iter()
        .map(|definition| {
            if !ids.insert(definition.id.clone()) {
                return Err("Duplicate case ID".into());
            }
            resolve_case(root, definition)
        })
        .collect()
}

pub fn resolve_case(root: &Path, definition: BehavioralCase) -> Result<LoadedCase, String> {
    definition.validate()?;
    let sources = definition
        .sources
        .iter()
        .map(|source| {
            let content = read_source(root, &source.path)?;
            section(&content, &source.heading)?;
            Ok(ResolvedSource {
                path: source.path.clone(),
                heading: source.heading.clone(),
                fingerprint: fingerprint(content),
            })
        })
        .collect::<Result<Vec<_>, String>>()?;
    let prompt = match &definition.prompt {
        Prompt::Inline { text } => text.clone(),
        Prompt::Trigger {
            trigger_file,
            query_index,
        } => load_trigger(root, trigger_file)?
            .queries
            .get(*query_index)
            .map(|q| q.query.clone())
            .unwrap_or_default(),
    };
    if prompt.is_empty() {
        return Err(format!("Unresolved prompt: {}", definition.id));
    }
    let fixture: Fixture = serde_json::from_str(&read_source(
        root,
        &format!("harness/evals/fixtures/{}.json", definition.fixture),
    )?)
    .map_err(|e| e.to_string())?;
    fixture.validate()?;
    if fixture.id != definition.fixture {
        return Err("Fixture identity mismatch".into());
    }
    Ok(LoadedCase {
        definition,
        sources,
        prompt_fingerprint: fingerprint(&prompt),
        prompt,
        fixture,
    })
}

#[cfg(test)]
#[path = "sources/tests.rs"]
mod tests;
